src/dependencies/__tests__/connect-test.js

```diff
@@ -137,6 +137,49 @@ describe('connect', () => {
     });
   });
 
+  describe('complex dependency results', () => {
+    it('prevents infinite loops with incorrectly diffed results', () => {
+      const simpleArraysStore = new StoreFactory({})
+        .defineGet(state => state)
+        .defineGetInitialState(() => [1, 2, 3, 4, 5])
+        .register(dispatcher);
+      const complexArraysStore = new StoreFactory({})
+        .defineGet(state => state)
+        .defineGetInitialState(() => [
+          { top: 'test', top2: [1, 2, 3], top3: { deep: 'test', deep2: 1234 } },
+          {
+            top: 'test2',
+            top2: [4, 5, 6],
+            top3: { deep: 'test2', deep2: 2468 },
+          },
+        ])
+        .register(dispatcher);
+      MockComponent = connect(
+        {
+          simpleArraysStore,
+          complexArraysStore,
+          merged: {
+            stores: [simpleArraysStore, complexArraysStore],
+            deref(__, depMap) {
+              console.log(depMap);
+              return (
+                depMap.simpleArraysStore &&
+                depMap.simpleArraysStore.concat(depMap.complexArraysStore)
+              );
+            },
+          },
+        },
+        dispatcher
+      )(BaseComponent);
+      let root;
+      act(() => {
+        root = mount(<MockComponent />);
+      });
+      // would not get here if array diffing wasn't working properly
+      expect(root.find(BaseComponent).exists()).toBe(true);
```

Comment:
Should there be a check that the deref isn't called more than 2 times? Or do the tests just crash otherwise?

Reply:
the tests just crash, React throws an exception if it detects a render loop (the heuristic is some arbitrary number of re-renders without any props/state changes)